Kernel Initializer issue

ananno opened the pull request:
<!--- 👉👉👉 Provide a general summary of your changes. 👈👈👈 -->
To fix the kernel initializer issue in function `conv2d` in `lmnet/lmnet/layers.py`.

## Motivation and Context
<!--- 💭💡💭 Why is this change required? What problem does it solve? 💭💡💭 -->
<!--- 🔗 If it fixes an open issue, please link to the issue here. 🔗 -->
The argument definition of `kernel_initializer` of the function `conv2d` in `lmnet/lmnet/layers/layers.py` has never been passed into `tf.layers.conv2d` function call resulting it not effective even if it is passed from the calling function. This change is necessary to pass the parameters in `tf.layers.conv2d` function by defining and passing arguments like `kernel_initializer` or `kernel_regularizer` etc.

## Description
<!--- 📝🎯📝 Describe your changes in detail. 📝🎯📝 -->
The `kernel_initializer` argument is defined in the function argument of `conv2d` but never passed in `tf.layers.conv2d` function in the code body. This results to not passing the kernel regularizer defined or passed by the calling function. 

The fix is very simple. By removing the argument definition in `conv2d` function in `lmnet/lmnet/layers/layers.py` file. The defined argument like `kernel_initializer` or any others supported by `tf.layers.conv2d` function can be automatically passed via `*args, **kwargs`.

Details discussion on this is documented in issue [here](https://github.com/blue-oil/blueoil/issues/362).

## How has this been tested?
<!--- 🙆👌🙆 Please describe in detail how you tested your changes. 🙆👌🙆 -->
<!--- Include details of your testing environment, details of your manual tests, snippet code or commands of your manual tests, table of experiments results metrics, tests ran to see how your change affects other areas of the code, etc. -->
<!--- If you have experiments report documents please link to here. -->
From the code it is obvious that the argument definition of `kernel_initializer` in `conv2d` function has never been used.

## Screenshots (if appropriate):
![image](https://user-images.githubusercontent.com/3903030/61017452-3a681180-a3ce-11e9-8168-4001adcc2a9e.png)

## Types of changes
<!--- 🏁🎌🏁 What types of changes does your code introduce? Put an `x` in all the boxes that apply: 🏁🎌🏁 -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature / Optimization (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- ✅✅✅ Go over all the following points, and put an `x` in all the boxes that apply. ✅✅✅ -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

<!--- TODO(wakisaka): After decide our code style, add this.
- [x] My code follows the code style of this project.
-->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

Joeper214 commented:
@ananno Please rebase this branch from the current `blueoil/master` branch. 

So that we can only see the commits from your branch.
Currently, there are `1,004` commits which many are unrelated to this PR.

ananno commented:
> @ananno Please rebase this branch from the current `blueoil/master` branch.
> 
> So that we can only see the commits you did on your branch.
> Currently, there are `1,004` commits which many are unrelated to this PR.

I've resolved the issue. Please check and confirm.
Thank you.

Joeper214 commented:
run test

iizukak commented:
@ananno 
Thank you for fixing the bug!!
As you comment in this PR, current lmnet implementation looks not pass `kernel_initializer` to TensorFlow.
So, currently, TensorFlow use default value `None`
https://www.tensorflow.org/api_docs/python/tf/layers/conv2d

And if your code will be merged, we use `xavier_initializer` as default (in your screenshot).

I thinks make `None` as default looks better.
How do you think?

iizukak commented:
@nlpng 
This PR effect many network.
It's good to be merge?

iizukak commented:
I checked TensorFlow use `glorot_uniform` when `kernel_initializer` is `None`.
So there is no problem about default initializer.

iizukak commented:
run test

Joeper214 commented:
@ananno Please fix failed tests as well.

ananno commented:
@tsawada San: Can you please do Readability review on this PR.
Thank you

ananno commented:
@iizukak San: You approved this PR but I think according to new policy you've to approve the ownership review. Please approve as ownership review.

Thank you

iizukak commented:
Please check CI status.

iizukak commented:
/ready

ananno commented:
@iizukak @tsawada San:
I though this would be a very simple fix (passing the `kernel_initializer` through `**kwargs` but as i investigate the test failure in `yolo_v2` training it expose another significant bug. 

The `kernel_initializer` parameter passing from darknet block was never passed into `conv2D` before. But when I fixed `conv2D` it turns out the parameter passing. Which means so far, all the network inherits yolov2, i.e. `lmfyolo`, didin't use `variance_scaling_initializer` but use the default normalized xavier initializer. This should cause performance difference in YoloV2 and all it's inheritor performance. 

As you already know for relu activation it was beneficial to use `variance_scaling_initializer` as kernel_initializer but the implement was wrong. Also in TF 1.15.0 the implementation was also changed.

Anyway, I've fixed both the issue and now all the tests are passed. I think it was a good finding. Please review one more time.

tsawada commented:
I think @tkng san knows better so I'll leave readability review to him.

iizukak commented:
@tkng 
I think we can merge this PR.

ananno commented:
@tkng San:
Can you please review one more time.

Thank you

ananno commented:
@iizukak San:
Can you please review one more time and if it is ok please give a OA and LGTM on this PR.

iizukak commented on lmnet/lmnet/layers/layers.py:
```diff
@@ -50,7 +50,6 @@ def conv2d(
     strides=1,
     padding="SAME",
     activation=tf.nn.relu,
-    kernel_initializer=tf.contrib.layers.xavier_initializer(),
```
@ananno Sorry, I have one question, before your change, we use `layers.xavier_initializer` and ignore `variance_scaling_initializer` by bug. 
Your fix change initializer `xavier_initializer` to `variance_scaling_initializer` and it's affect training accuracy?  🤔

ananno replied:
I didn't trained with both. But theoretically it would better perform with `variance_scaling_initializer` as it works better with `ReLU` activation whereas `xavier_initializer` works better with sigmoid activation.
Ref. Xavier : http://proceedings.mlr.press/v9/glorot10a/glorot10a.pdf
Ref. Variance Scaling : https://arxiv.org/pdf/1502.01852v1.pdf

iizukak replied:
I want @tkng san's opinion about this.

ananno replied:
Moreover, before we didn't pass `kernel_initializer` into the `tf.layers.conv2d` function anyway, so it would use the default value always and our selection through config was not being applied.

iizukak replied:
I see.

ananno commented:
@tsawada San:
Can you please review this PR for Readability Approval.

tkng commented on lmnet/lmnet/blocks.py:
```diff
@@ -40,7 +40,7 @@ def activation(x): return tf.nn.leaky_relu(x, alpha=0.1, name="leaky_relu")
         conv = conv2d("conv", inputs, filters=filters, kernel_size=kernel_size,
                       activation=None, use_bias=False, data_format=channel_data_format,
-                      kernel_initializer=tf.contrib.layers.variance_scaling_initializer(),)  # he initializer
+                      kernel_initializer=tf.variance_scaling_initializer,)  # he initializer
```
According to the reference manual, `tf.contrib.layers.variance_scaling_initializer` is he initializer (MSRA initializer), but `tf.variance_scaling_initializer` is not. (Latter doesn't have an explicit short name, Caffe's default.)

- https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/contrib/layers/variance_scaling_initializer
- https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/initializers/variance_scaling

I think it's better to keep `tf.contrib.layers.variance_scaling_initializer()`, since he initializer is major than `tf.variance_scaling_initializer`'s default setting.

ananno replied:
I see. In that case we can keep it for now. But it will be deprecated in TF2.0, until than we can use it.
 
![Screenshot from 2020-01-29 16-35-16](https://user-images.githubusercontent.com/3903030/73336665-c03da280-42b5-11ea-845d-ac5d7dca3b47.png)

ananno commented:
> Please revert this change, pull request should not include any cosmetic change without reason.
> 
> (I meant this line
> 
> https://github.com/blue-oil/blueoil/blob/984cd163bd26f80efb6989fea29ca94824615603/lmnet/lmnet/blocks.py#L43
> 
> )

@tkng San:
Thank you for the explanation. I've revert the cosmetic change. Please review one more time.

tkng commented:
Readability Approval
Ownership Approval

tkng commented:
/ready

ananno commented:
/ready

ananno commented:
/ready
